File: Mobile App/AndroidAppScan/app/src/main/java/com/tool/scanqrasset/ui/ScanQRcodeActivity.java
package com.tool.scanqrasset.ui;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.cardview.widget.CardView;

import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.vision.barcode.Barcode;
import com.marcinorlowski.fonty.Fonty;
import com.tool.scanqrasset.MainApplication;
import com.tool.scanqrasset.R;
import com.tool.scanqrasset.barcode.BarcodeCaptureActivity;
import com.tool.scanqrasset.utils.ClipboardUtils;
import com.tool.scanqrasset.utils.StoreDataHelper;

import java.net.URL;

public class ScanQRcodeActivity extends AppCompatActivity {
    TextView lblResult;
    TextView tvTimeLefts;
    Button btnBuy;
    ImageView imgCopy, imgOpenLink, imgShare, imgBack, imgPremium;
    CardView bgScan;
    int BARCODE_READER_REQUEST_CODE = 1;

    ////////////////////////////////////////////////////////////////////////////////////////////////

    private Toolbar mActivityActionBarToolbar;

    @Override
    public void setSupportActionBar(@Nullable Toolbar toolbar) {
        super.setSupportActionBar(toolbar);
        mActivityActionBarToolbar = toolbar;
    }

    @Override
    public void setTitle(CharSequence title) {
        ActionBar ab = getSupportActionBar();
        if (ab != null) {
            ab.setTitle(title);
            Fonty.setFonts(mActivityActionBarToolbar);
        }
    }

    public void setSubtitle(CharSequence subtitle) {
        ActionBar ab = getSupportActionBar();
        if (ab != null) {
            ab.setSubtitle(subtitle);
            Fonty.setFonts(mActivityActionBarToolbar);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.screen_scan_qr_code);


        lblResult = findViewById(R.id.lblResult);
        tvTimeLefts = findViewById(R.id.tvTimeLefts);
        btnBuy = findViewById(R.id.btnBuy);
        imgCopy = findViewById(R.id.imgCopy);
        imgShare = findViewById(R.id.imgShare);
        imgOpenLink = findViewById(R.id.imgOpen);
        bgScan = findViewById(R.id.btnScan);
        imgBack = findViewById(R.id.btnBack);
        imgPremium = findViewById(R.id.imgPremium);

//        getSupportActionBar().setDisplayHomeAsUpEnabled(false);


        btnBuy.setOnClickListener(v -> {
            Intent intent = new Intent(this, SubsActivity.class);
            ScanQRcodeActivity.this.startActivity(intent);
        });

        imgCopy.setOnClickListener(v -> {
            copyText();
        });

        imgOpenLink.setOnClickListener(v -> {
            openLink();
        });

        imgShare.setOnClickListener(v -> {
            shareText();
        });

        bgScan.setOnClickListener(v -> {
            scanQr();
        });

        imgBack.setOnClickListener(v -> {
            finish();
        });

        imgPremium.setOnClickListener(v -> {
            int timeLefts = StoreDataHelper.getInstance().getDayLefts();
            String dateLeft;
            if (timeLefts > 1) {
                dateLeft = timeLefts + " days";
            } else {
                dateLeft = "0 day";
            }
            String content = "You are using Premium version\nYour time lefts: " + dateLeft;

            Toast.makeText(this, content, Toast.LENGTH_SHORT).show();

//            Intent intent = new Intent(this, SubsActivity.class);
//            startActivity(intent);
        });

        Fonty.setFonts(this);

//        setTitle("QR Scan");

        try {
            if (getIntent() != null) {
                Barcode barcode = getIntent().getParcelableExtra("Barcode");
                if (barcode != null) {
                    lblResult.setText(barcode.displayValue);
                }
            }
        } catch (Exception ex) {
            Toast.makeText(this, ex.getMessage(), Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        int dateLeft = StoreDataHelper.getInstance().getDayLefts();

        tvTimeLefts.setText("Your date lefts: " + (dateLeft > 0 ? dateLeft : 0));
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == BARCODE_READER_REQUEST_CODE) {
            if (resultCode == CommonStatusCodes.SUCCESS) {
                try {
                    if (data != null) {
                        Barcode barcode = data.getParcelableExtra("Barcode");
                        if (barcode != null) {
                            lblResult.setText(barcode.displayValue);
                        }
                    }
                } catch (Exception ex) {
                    Toast.makeText(this, ex.getMessage(), Toast.LENGTH_LONG).show();
                }
            }
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    private void scanQr() {
        int timeLefts = StoreDataHelper.getInstance().getDayLefts();
        Intent intent;
        if (timeLefts > 0) {
            intent = new Intent(this, BarcodeCaptureActivity.class);
        } else {
            intent = new Intent(this, SubsActivity.class);
        }
        startActivity(intent);

    }

    private void copyText() {
        if (lblResult.getText() != null) {
            if (ClipboardUtils.get(MainApplication.getInstance()).copy(lblResult.getText().toString())) {
                Toast.makeText(this, "Copy Successfully", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(this, "Copy Failed", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void shareText() {
        if (lblResult.getText() != null) {
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            String content = lblResult.getText().toString();
            shareIntent.putExtra(Intent.EXTRA_TEXT, content);
            startActivity(Intent.createChooser(shareIntent, "Share via").addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
        }
    }

    private void openLink() {
        if (lblResult.getText() != null) {
            String text = lblResult.getText().toString();

            try {
                URL u = new URL(text);
                u.toURI();

                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(text));
                startActivity(browserIntent);
            } catch (Exception e) {
                Toast.makeText(this, "Invalid Url", Toast.LENGTH_SHORT).show();
            }

        }
    }
}

File: Mobile App/AndroidAppScan/app/src/main/java/com/tool/scanqrasset/ui/ScanQRcodeActivity.kt
package com.tool.scanqrasset.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.cardview.widget.CardView
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.vision.barcode.Barcode
import com.marcinorlowski.fonty.Fonty
import com.tool.scanqrasset.MainApplication
import com.tool.scanqrasset.R
import com.tool.scanqrasset.barcode.BarcodeCaptureActivity
import com.tool.scanqrasset.databinding.ScreenScanQrCodeBinding
import com.tool.scanqrasset.inf.FetchApiAction
import com.tool.scanqrasset.model.Asset
import com.tool.scanqrasset.utils.ClipboardUtils
import com.tool.scanqrasset.utils.FetchApiHelper
import com.tool.scanqrasset.utils.StoreDataHelper
import kotlinx.coroutines.CoroutineExceptionHandler
import java.net.URL

class ScanQRcodeActivity : AppCompatActivity(), FetchApiAction {
    var searchView: SearchView? = null
    var tvTimeLefts: TextView? = null
    var btnBuy: Button? = null
    var imgCopy: ImageView? = null
    var imgOpenLink: ImageView? = null
    var imgShare: ImageView? = null
    var imgBack: ImageView? = null
    var imgPremium: ImageView? = null
    var imgScan: ImageView? = null
    var bgScan: CardView? = null
    private val BARCODE_READER_REQUEST_CODE = 1
    private var binding: ScreenScanQrCodeBinding? = null

    ////////////////////////////////////////////////////////////////////////////////////////////////
    private var mActivityActionBarToolbar: Toolbar? = null
    override fun setSupportActionBar(toolbar: Toolbar?) {
        super.setSupportActionBar(toolbar)
        mActivityActionBarToolbar = toolbar
    }

    override fun setTitle(title: CharSequence) {
        val ab = supportActionBar
        if (ab != null) {
            ab.title = title
            Fonty.setFonts(mActivityActionBarToolbar)
        }
    }

    fun setSubtitle(subtitle: CharSequence?) {
        val ab = supportActionBar
        if (ab != null) {
            ab.subtitle = subtitle
            Fonty.setFonts(mActivityActionBarToolbar)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ScreenScanQrCodeBinding.inflate(layoutInflater)
        val view: View = binding!!.root
        setContentView(view)
        searchView = findViewById(R.id.searchView)
        tvTimeLefts = findViewById(R.id.tvTimeLefts)
        btnBuy = findViewById(R.id.btnBuy)
        imgCopy = findViewById(R.id.imgCopy)
        imgShare = findViewById(R.id.imgShare)
        imgOpenLink = findViewById(R.id.imgOpen)
        bgScan = findViewById(R.id.btnScan)
        imgBack = findViewById(R.id.btnBack)
        imgPremium = findViewById(R.id.imgPremium)
        imgScan = findViewById(R.id.imgScan)

//        getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        imgScan!!.setOnClickListener(View.OnClickListener { v: View? ->
            val intent = Intent(this, BarcodeCaptureActivity::class.java)
            this@ScanQRcodeActivity.startActivity(intent)
        })
        imgCopy!!.setOnClickListener(View.OnClickListener { v: View? -> copyText() })
        imgOpenLink!!.setOnClickListener(View.OnClickListener { v: View? -> openLink() })
        imgShare!!.setOnClickListener(View.OnClickListener { v: View? -> shareText() })

//        bgScan.setOnClickListener(v -> {
//            scanQr();
//        });
        imgBack!!.setOnClickListener(View.OnClickListener { v: View? -> finish() })
        imgPremium!!.setOnClickListener(View.OnClickListener { v: View? ->
            val timeLefts = StoreDataHelper.getInstance().getDayLefts()
            val dateLeft: String
            dateLeft = if (timeLefts > 1) {
                "$timeLefts days"
            } else {
                "0 day"
            }
            val content = "You are using Premium version\nYour time lefts: $dateLeft"
            Toast.makeText(this, content, Toast.LENGTH_SHORT).show()
        })

        // search view
        searchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
//                Toast.makeText(getBaseContext(), query, Toast.LENGTH_SHORT).show();
                searchAssetById(query)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })
        Fonty.setFonts(this)

        binding!!.bgInfo.visibility = View.GONE

//        setTitle("QR Scan");
        try {
            if (intent != null) {
                val barcode = intent.getParcelableExtra<Barcode>("Barcode")
                if (barcode != null) {
                    binding!!.searchView.setQuery(barcode.displayValue, true)
                }
            }
        } catch (ex: Exception) {
            Toast.makeText(this, ex.message, Toast.LENGTH_LONG).show()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun searchAssetById(id: String) {
        FetchApiHelper.getInstance()
            .fetchAssetsApi(id, this, CoroutineExceptionHandler { _, throwable ->
                Toast.makeText(
                    MainApplication.getInstance().applicationContext,
                    throwable.message,
                    Toast.LENGTH_SHORT
                )
                    .show()

                binding!!.bgInfo.visibility = View.GONE

                binding!!.bgMsg.visibility = View.VISIBLE
                binding!!.lbMsg.text = "Not found ID: $id";
            })
    }

    override fun onResume() {
        super.onResume()
        val dateLeft = StoreDataHelper.getInstance().getDayLefts()
        tvTimeLefts!!.text = "Your date lefts: " + if (dateLeft > 0) dateLeft else 0
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == BARCODE_READER_REQUEST_CODE) {
            if (resultCode == CommonStatusCodes.SUCCESS) {
                try {
                    if (data != null) {
                        val barcode = data.getParcelableExtra<Barcode>("Barcode")
                        if (barcode != null) {
                            searchView!!.setQuery(barcode.displayValue, true)
                        }
                    }
                } catch (ex: Exception) {
                    Toast.makeText(this, ex.message, Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    private fun scanQr() {
        val timeLefts = StoreDataHelper.getInstance().getDayLefts()
        val intent: Intent = if (timeLefts > 0) {
            Intent(this, BarcodeCaptureActivity::class.java)
        } else {
            Intent(this, SubsActivity::class.java)
        }
        startActivity(intent)
    }

    private fun copyText() {
        if (searchView!!.query != null) {
            if (ClipboardUtils.get(MainApplication.getInstance())
                    .copy(searchView!!.query.toString())
            ) {
                Toast.makeText(this, "Copy Successfully", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, "Copy Failed", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun shareText() {
        if (searchView!!.query != null) {
            val shareIntent = Intent(Intent.ACTION_SEND)
            shareIntent.type = "text/plain"
            val content = searchView!!.query.toString()
            shareIntent.putExtra(Intent.EXTRA_TEXT, content)
            startActivity(
                Intent.createChooser(shareIntent, "Share via")
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            )
        }
    }

    private fun openLink() {
        if (searchView!!.query != null) {
            val text = searchView!!.query.toString()
            try {
                val u = URL(text)
                u.toURI()
                val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(text))
                startActivity(browserIntent)
            } catch (e: Exception) {
                Toast.makeText(this, "Invalid Url", Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun doAction(value: Any) {
        // Only runs if there is a view that is currently focused
        this.currentFocus?.let { view ->
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
            imm?.hideSoftInputFromWindow(view.windowToken, 0)
        }

        val asset  = value as Asset

        binding!!.bgInfo.visibility = View.VISIBLE
        binding!!.bgMsg.visibility = View.GONE

        binding!!.tvPartId.text = asset.partId
        binding!!.tvDescription.text = asset.description
        binding!!.tvUseFor.text = asset.useFor
        binding!!.tvStorageBin.text = asset.storageBin
        binding!!.tvStockAmount.text = asset.stockAmount.toString()
        binding!!.tvSearch.text = asset.search
        binding!!.tvBrand.text = asset.brand
        binding!!.cbLocal.isChecked = asset.local
        binding!!.cbImport.isChecked = asset.import
        binding!!.tvNote.text = asset.note
        binding!!.tvAlp.text = asset.alp
        binding!!.tvSupplier.text = asset.supplier
    }
}